### DISCUSSION TITLE
With the 7.0 RC Java API changes, how do I run against something other than a file?

### DISCUSSION THREAD

**Discussion by SCWells72** [pmd/pmd#4595] [Q&A] [answered]
I currently integrate PMD for real-time analysis into my product, and it operates against the latest in-memory version of a file rather than the on-disk copy. I do this by creating a `ReaderDataSource` and passing that to `PMD.processFiles()`. With the change to `PmdAnalysis`, I can't see a way to replicate this behavior. I'm sure I'm just missing something, but after a bit of digging, I'm just not seeing it. Basically I'm trying to figure out how to accomplish the following in 7.0+:

```java
DataSource fileBodyDataSource = new ReaderDataSource(new StringReader(fileBody), filePath);
List<DataSource> dataSources = Collections.singletonList(fileBodyDataSource);
Report report = PMD.processFiles(configuration, ruleSets, dataSources, Collections.emptyList());
```

with a specific focus on the `dataSources` part. I've already seen how to add `ruleSets` to the `configuration`.

Thanks in advance for any advice!

**Comment by oowekyala** [comment, marked as answer]
DataSource has been replaced by [TextFile](https://docs.pmd-code.org/apidocs/pmd-core/7.0.0-SNAPSHOT/net/sourceforge/pmd/lang/document/TextFile.html) in PMD 7. You can implement the interface yourself, but there are premade factory methods on the interface if your file contents come from a reader (or in this case, a string).

(Note that you need to pass in the language version of the file when creating it)
```java
TextFile file = TextFile.forCharSeq(fileBody, FileId.fromPathLikeString(filePath), languageVersion);
try (PmdAnalysis pmd = PmdAnalysis.create(configuration)) {
   pmd.files().addFile(file);
   // also add your rulesets and other stuff if they're not in the configuration
   Report report = pmd.performAnalysisAndCollectReport();
}
```

**Reply by SCWells72** [reply]
Perfect! Thank you so much for the quick and detailed response.

**Comment by SCWells72** [comment]
I apologize for the long-delayed follow-up. So I changed this (6.55.0 and earlier):

```java
DataSource fileBodyDataSource = new ReaderDataSource(new StringReader(fileBody), filePath);
List<DataSource> dataSources = Collections.singletonList(fileBodyDataSource);
Report report = PMD.processFiles(configuration, ruleSets, dataSources, Collections.emptyList());
```

to this (7.0.0-rc3):

```java
FileId fileId = FileId.fromPathLikeString(filePath);
PmdAnalysis pmdAnalysis = PmdAnalysis.create(configuration);
pmdAnalysis.addRuleSets(ruleSets);
LanguageVersion languageVersion = configuration.getLanguageVersionOfFile(PathUtil.getFileName(filePath));
TextFile textFile = TextFile.forCharSeq(fileBody, fileId, languageVersion);
pmdAnalysis.files().addFile(textFile);
Report report = pmdAnalysis.performAnalysisAndCollectReport();
```

and while the latter executes, it yields an empty report. I imagine I'm just missing something silly, but I can't seem to figure out what that might be at a glance. Any thoughts are greatly appreciated.

**Reply by oowekyala** [reply]
Can you post the debug output? (`--debug`)

**Reply by SCWells72** [reply]
I'm happy to get what you want, but this is all being done directly via the Java API. How would I retrieve what you're wanting when it's not hooked up to `stdout` and `stderr`?

**Reply by SCWells72** [reply]
FYI, I figured this out and it was in fact an issue on my side. I've resolved that. I'm going to investigate the other question below next, specifically whether it's possible to have a single implementation that works against both the 6.x and 7.x APIs. Any thoughts you might have on that, specifically if it's not possible (without extensive reflection, at least).

**Comment by SCWells72** [comment]
And once this one is figured out, I'll have another question about whether it's possible to write a client app against the PMD Java API that works against both 6.55.0 and 7.x distributions. For context, I've integrated PMD into a JetBrains IDE plugin for real-time analysis where the PMD jars aren't distributed with my plugin; instead it uses a dynamic classloader against an identified distribution directory and ruleset file. This has been working very, very well for quite a while, and now I'm trying to update it to work against 7.0.0-rc3. In the end, ideally it would work against both 6.x for those who haven't yet updated and 7.x for those who have. Not sure if that's going to be possible, but I guess the first step is getting a report from 7.x...

**Reply by SCWells72** [reply]
@oowekyala, it looks like 6.55.0 doesn't have `TextFile.forCharSeq()`, at least not in an accessible form, nor does it have `FileId`. Based on that, are you aware of a way to express the equivalent of the following in both 6.x and 7.x?

```java
FileId fileId = FileId.fromPathLikeString(filePath);
...
TextFile textFile = TextFile.forCharSeq(fileBody, fileId, languageVersion);
```

I guess worst case scenario, I could use reflection to get from the file body to the `Report` according to the underlying version, but my guess is that that's going to get pretty ugly. Hopefully there's some common interface for this between 6.x and 7.x that I'm just not seeing right now.

**Reply by SCWells72** [reply]
Okay, I also got this working via some reflection hackery (simplified version without more extensive error handling/reporting):

```java
private static TextFile getTextFile(String filePath, String fileBody, LanguageVersion languageVersion) {
    TextFile textFile = null;

    try {
        // If we have access to FileId, use it (7.x+)
        Class<?> fileIdClass = null;
        try { fileIdClass = Class.forName("net.sourceforge.pmd.lang.document.FileId"); } catch (ClassNotFoundException e) {/* Ignore */}
        if (fileIdClass != null) {
            Method fromPathLikeStringMethod = ReflectionUtil.getMethod(fileIdClass, "fromPathLikeString", String.class);
            if (fromPathLikeStringMethod != null) {
                Object fileIdCandidate = fromPathLikeStringMethod.invoke(null, filePath);
                if ((fileIdCandidate != null) && fileIdClass.isAssignableFrom(fileIdCandidate.getClass())) {
                    Method forCharSeqMethod = ReflectionUtil.getMethod(TextFile.class, "forCharSeq", CharSequence.class, fileIdClass, LanguageVersion.class);
                    if (forCharSeqMethod != null) {
                        Object textFileCandidate = forCharSeqMethod.invoke(null, fileBody, fileIdCandidate, languageVersion);
                        if (textFileCandidate instanceof TextFile) {
                            textFile = (TextFile) textFileCandidate;
                        }
                    }
                }
            }
        }
        // Otherwise use StringTextFile (6.x)
        else
        {
            Class<?> stringTextFileClass = Class.forName("net.sourceforge.pmd.lang.document.StringTextFile");
            if (stringTextFileClass != null) {
                Constructor<?> stringTextFileConstructor = stringTextFileClass.getDeclaredConstructor(String.class, String.class, String.class, LanguageVersion.class);
                if (stringTextFileConstructor != null) {
                    stringTextFileConstructor.setAccessible(true);
                    String fileName = PathUtil.getFileName(filePath);
                    Object textFileCandidate = stringTextFileConstructor.newInstance(fileBody, filePath, fileName, languageVersion);
                    if (textFileCandidate instanceof TextFile) {
                        textFile = (TextFile) textFileCandidate;
                    }
                }
            }
        }
    } catch (Exception e) {
        LOG.warn("Failed to create a TextFile for '" + filePath + "'.", e);
    }

    return textFile;
}
```

Remember that this is being executed in the context of a custom classloader with the jars from the respective distribution, so I can actually freely switch back and forth between 6.55.0 and 7.0.0-rc3 and have them both work properly in the same process.

Obviously not ideal, but I have a big TODO marker saying that this can be removed and replaced with a static version of the 7.x stuff once my plugin no longer supports 6.x. I'll likely wait a while for 7.x to gel and get traction before doing that, though.

If you have thoughts on how to do this better, I'm absolutely open to them.

**Reply by adangel** [reply]
> specifically whether it's possible to have a single implementation that works against both the 6.x and 7.x APIs

A similar approach has been started with #4609 - which basically provides a 6.x compatible entry point to PMD 7. Maybe that could be done this other way round as well. I was thinking about something like a "pmd-compat" module, which you can add as an additional dependency...

**Reply by SCWells72** [reply]
Hi. The workaround I described previously continues to work just fine. I did have to add a few additional jars to the custom classloader's classpath to keep things working in 7.0.0-rc4, but no further code changes were needed. I'm fine with the aforementioned workaround for the period when I support both 6.x and 7.x. At least for me, there's no need for a compatibility facade. I do appreciate the consideration, though.
